Rework register representation

GMMan opened the pull request:
Mostly for awareness, there's a lot of other bits and pieces still floating in here.

To help sidestep the issue with incorrect CS, I took a page from the ARM SLEIGH processor and separated out all of the fields from SR into separate registers. This reduced the amount of extraneous statements generated by the decompiler. PC is represented by a 3-byte register, with the actual PC and CS overlaid on top of it. Hopefully that will help resolve CS not updating correctly, unless I'm grossly misunderstanding how the instruction pointer works in SLEIGH. SR is now a "virtual" register, only packed and unpacked when it's accessed, which while there are a lot of instructions that could access it, it is thankfully done fairly infrequently in code (that I'm looking at, at least).

GMMan committed:
Make registers discrete

GMMan committed:
Update language name

GMMan committed:
Add missing Direct6 memory assignment, remove invalid "0:" prefix

GMMan committed:
Fix prototype stack shift/pop

GMMan committed:
Omit SR packing during call

SR does not change as a result of a call, so it's OK that we don't pack
the real value of it. Callees shouldn't need to access the saved SR
typically.

GMMan committed:
Revert to simpler push macro

GMMan commented:
Per comment [here](https://github.com/NationalSecurityAgency/ghidra/issues/2900#issuecomment-813695832), packing SR is not strictly necessary if it won't change as a result of a call (which it generally shouldn't), so I've removed `PackSR` from `call` instructions to prevent extraneous output.

GMMan committed:
Simplify conditional jumps

GMMan committed:
Fix C and S flags handling

GMMan commented:
Note that `cmp` *does not* read the carry flag. The previous code resulted in bizarre decompilation dependent on previous subtract with carry results, turning a simple compare and branch into something nonsensical.

GMMan committed:
Add function start patterns

GMMan committed:
Fix return address pushing for call

GMMan committed:
Add memory BITOP A16 support

GMMan committed:
Add new token fields

SamuelWAnderson45 commented:
Oooh thanks for the help. I've been dropped off the map for a while, but I'll take a closer look next week.
I almost made `cs` part of `pc_full`, but I wasn't sure how to handle `sr`. I really like what you've done with the Pack/Unpack thing at first glance.
As you can probably tell, I didn't have access to the Programmer's Guide when I made this; I only just found it a couple months ago. So plenty of silly bugs and room for improvement :)

SamuelWAnderson45 commented:
So I compile with the `-l` flag to detect pattern conflicts and I get this
```
ERROR unsp.slaspec:348: Constructor patterns cannot be distinguished: 
   InstructionPattern{00...... 0000...0} table "instruction" constructor from unsp.slaspec:348
   InstructionPattern{00...... ....1110} table "instruction" constructor from unsp.slaspec:224 (SleighCompile)  
ERROR unsp.slaspec:224: Constructor patterns cannot be distinguished: 
   InstructionPattern{00...... 0000...0} table "instruction" constructor from unsp.slaspec:348
   InstructionPattern{00...... ....1110} table "instruction" constructor from unsp.slaspec:224 (SleighCompile)
```
Basically its ambiguous how to parse some bitstrings. The docs explain it best: https://ghidra.re/courses/languages/html/sleigh_constructors.html#sleigh_tables

>  The bit patterns for constructors under a single table must be built so that a constructor can be uniquely determined in context. The above example shows the easiest way to accomplish this. The two sets of instruction encodings, which match one or the other of the two addrmode constraints, are disjoint. In general, if each constructor has a set of instruction encodings associated with it, and if the sets for any two constructors are disjoint, then no two constructors can match at the same time.
> It is possible for two sets to intersect, if one of the two sets properly contains the other. In this situation, the constructor corresponding to the smaller (contained) set is considered a special case of the other constructor. If an instruction encoding matches the special case, that constructor is used to define the symbol, even though the other constructor will also match. If the special case does not match but the other more general constructor does, then the general constructor is used to define the symbol.

Unfortunately, its not theoretical, I've been bitten by this in the past. And it doesn't always seem to be detected by the `-l` flag, but it helps. One particularly nasty variant is it _appears_ (I might be wrong) that the SLEIGH parser only evaluates the first matching constructor (based on the root table constraints). If that fails to parse (because constraints on the subtables don't match), then it gives up.

I know you can get away with it to a degree, by messing with the order the constructors are defined in the spec, but relying on that feels like a bad idea.

GMMan commented:
That's interesting, I did not know about that option. I'll see if I can tighten the patterns up more (I assumed they were set up with all the instruction class constraints already).

SamuelWAnderson45 commented:
Cool!
Here's all the flags I use `-u -l -t -c -f -e -x`. They all enable various warnings (a couple of which are mostly ignorable, but it's nice to have them).
Yeah I was taking advantage of the "special case" thing from the docs so when you refactored some constraints into subtables it made it no longer a "special case".

GMMan committed:
Fix return register combination order

GMMan committed:
Fix ridiculous far pointers

Turns out arguments to segmentop are left to right, not low order to
high order

SamuelWAnderson45 commented:
I know its been two years, but I'll probably merge this. Your changes are some pretty significant improvements.
Let me know if you have any thoughts @GMMan otherwise i'll merge this in a few days.
